### PR TITLE
MAINT: version bounds for 1.12.0rc1

### PR DESCRIPTION
**Pull request by tylerjereddy**
* draft the release version pins in (we don't upper bound Python anymore because of `poetry`, and NumPy `2.0.0` support is likely reserved for the next SciPy, `1.13.0`)

* ~point wheel builds at stable release NumPy (I think...), which I think should be sufficiently compatible because of:
https://numpy.org/doc/stable/release/1.25.0-notes.html#compiling-against-the-numpy-c-api-is-now-backwards-compatible-by-default~

* additional shims/corrections are almost certain to be necessary, but I'll start probing the wheel builds now

* a few release notes fixups I caught in the rendered version

Backports included so far:

1. gh-19693
2. gh-19690
3. gh-19686
4. gh-19677
5. gh-19691
6. gh-19702

[wheel build]

### CODE REVIEW

**Comment by tylerjereddy** [comment]
Correction: we're still using the old NumPy pins for wheel builds, and not quite sure we want to switch to testing the new thing where `1.25.x` has improved backwards compat.

Also, I opened this PR against the wrong base branch, correct that--need to be careful when multi-tasking...

**Comment by tylerjereddy** [comment]
I can reproduce the Python `3.12` wheel build failure locally if I uninstall `setuptools` and use `pythran` `0.14.0` (the latest stable release). The problem goes away if I install `setuptools` (presumably as a provider of `distutils`).

There is no improvement if I use the latest `master` branch of `pythran` locally, and uninstall `setuptools` after it pulls it in. Pythran seems to require `setuptools` with Python `3.12`. In the context of an isolated wheel build, can we expect something reasonable to happen via the transitive requirement for `setuptools` (should/could Pythran be responsible for expressing that in a way we can detect)? Or do we need a shim to `pyproject.toml` on our end? Or will we need to bite the bullet and remove build isolation to do something custom?

@rgommers @henryiii thoughts?

**Comment by tylerjereddy** [comment]
I should cc @h-vetinari here as well.

**Comment by henryiii** [comment]
Pythran doesn’t have a pyproject.toml. So it doesn’t declare its build dependencies. Ideally, they could add one which would fix the problem.

**Comment by tylerjereddy** [comment]
@serge-sans-paille when is the next Pythran release scheduled? Are you planning to add `pyproject.toml` per above?

**Comment by serge-sans-paille** [comment]
> @serge-sans-paille when is the next Pythran release scheduled? Are you planning to add `pyproject.toml` per above?

I can certainly do that. I have a few changes related to python 3.12 on top of latest release so that would make a good 'portability' release.

File: pyproject.toml [hunk 1]
```diff
@@ -50,11 +50,11 @@ maintainers = [
 # Note: Python and NumPy upper version bounds should be set correctly in
 # release branches, see:
 #     https://scipy.github.io/devdocs/dev/core-dev/index.html#version-ranges-for-numpy-and-other-dependencies
-requires-python = ">=3.9"
+requires-python = ">=3.9" # NOTE: we've stopped placing upper bounds here because of i.e., poetry
```
Comment (rgommers) [inline pyproject.toml:53]: I'd change this to: `# newest support Python version: 3.12`

File: meson.build [hunk 2]
```diff
@@ -53,6 +53,11 @@ endif
 if not cy.version().version_compare('>=0.29.35')
   error('SciPy requires Cython >= 0.29.35')
 endif
+# NOTE: Cython tends to do beta releases
+# that may cause this check to choke
```
Comment (rgommers) [inline meson.build:57]: If that's a risk, let's not add this? The constraint in `pyproject.toml` should be okay.

**Comment by rgommers** [comment]
Thanks Tyler. Overall looks good. +1 for waiting for `pythran` to fix its `setuptools` issue rather than try to reshuffle our wheel build scripts.

**Comment by serge-sans-paille** [comment]
https://github.com/serge-sans-paille/pythran/pull/2162 << Can you check that this work as expected wrt. scipy?

**Comment by henryiii** [comment]
FYI, you don't need to move the _build configuration_ to pyproject.toml. You only need to add a _build-system configuration_ (which is a totally different thing, and currently missing, hence the bug) to a new pyproject.toml file. If you then wanted to also move build config, that would be a good second step, but I think it would be easier in two steps, would hopefully change less than 2,000 files in the first step that way.

**Comment by serge-sans-paille** [comment]
> FYI, you don't need to move the _build configuration_ to pyproject.toml. You only need to add a _build-system configuration_ (which is a totally different thing, and currently missing, hence the bug) to a new pyproject.toml file. If you then wanted to also move build config, that would be a good second step, but I think it would be easier in two steps, would hopefully change less than 2,000 files in the first step that way.

oh, I'm actually pretty happy to get rid of the clutter from current `setup.py` :-)

**Comment by tylerjereddy** [comment]
I've updated to address the review comments, and I've also added all the backports I'm aware of (summarized in the original comment above now).

I'll let regular CI (minus Cirrus) flush on this push, but I won't try flushing the wheels until the new Pythran comes out.

One other thought I had while doing backports is that I backported the fix for ARM Mac + NumPy `main`, but haven't backported any of the patches for Windows int shims + NumPy `main`. If we get all those done before the first RC I could consider doing that maybe, but probably not critical if we do a `1.13.0` separately for NumPy `2.0.0` support.

**Comment by tylerjereddy** [comment]
Perhaps "note to self," `version = "1.12.0.dev0"` in `pyproject.toml` seems like something I might forget to tweak (not sure how important), but looking at my usual release steps I've recently just been mutating `meson.build` and `tools/version_utils.py`. Maybe not crucial in RC phase, but I suspect I should at least try to remove the `.dev0` for final release (the `version` field doesn't exist on the `maintenance/1.11.x` branch of previous release).

**Comment by h-vetinari** [comment]
Sorry for the late backport candidate, but I think we should add #19702 due to the current inconsistencies (introduced in the 1.12 cycle) between solvers in `scipy.sparse.linalg.*`, which that PR fixes.

**Comment by tylerjereddy** [comment]
Ok, if it gets reviewed/approved on time I'll backport it. At the moment I think we're waiting for a new Pythran release, though I could eventually decide to disable build isolation and do an RC earlier. I won't do that today, but could consider it in a few days if we get held up a lot.

**Comment by h-vetinari** [comment]
Thanks a lot. If necessary, I think the PR could also land between rc1 and GA, since it just introduces a few new deprecations, and doesn't influence the build or other lower-level things.

**Comment by h-vetinari** [comment]
Hey @tylerjereddy, https://github.com/scipy/scipy/pull/19702 now got merged but (alas!) not squashed. If it helps, I'm happy to prepare a squashed commit on top of this PR (either as a PR against your fork or as a direct push), or similarly, I can rebase the commits in question on top of this branch. Please let me know if you need anything from my side to make this easier for yourself.

**Comment by tylerjereddy** [comment]
> I'm happy to prepare a squashed commit on top of this PR

I fused the commits together manually, and this PR is now updated to include all merged backport PRs at the time of writing.

I'm also now inclined to consider temporary shims to the wheel builds for RC1 to allow testing to happen until the new Pythran gets released, so I may try to prepare such changes today, provided they can be done in a single easy-to-revert commit at the RC2 stage.

**Comment by tylerjereddy** [comment]
A test wheel build has been initiated with a one-line patch to temporarily use `pythran` `master` branch (fixed the issue locally with `3.12`). If the CI is all green with that, I think it will be release manager discretion to release RC1 in that state, then revert this simple diff in RC2, so we get some testing sooner than later:

```diff
--- a/pyproject.toml
+++ b/pyproject.toml
@@ -13,7 +13,7 @@ requires = [
     "meson-python>=0.15.0,<0.16.0", # working with 0.15.x series at branch time
     "Cython>=0.29.35,!=3.0.3,<3.1.0",  # when updating version, also update check in meson.build
     "pybind11>=2.10.4,<2.12.0", # working with 2.11.x series at branch time
-    "pythran>=0.14.0,<0.15.0", # working with 0.14.x series at branch time
+    "pythran@git+https://github.com/serge-sans-paille/pythran", # TODO: pin back to stable release when ready
```

**Comment by rgommers** [comment]
👍🏼 that strategy seems reasonable to me.

**Comment by tylerjereddy** [comment]
Ciruss ARM Linux timed out, I restarted that one failure. It looks like the other ARM Linux job was also close to the limit. Anyway, not a big deal if I need to restart those jobs during the release process, but before merging here I should check that restarting generally fixes the problem. Longer-term we may want to check on the performance of those jobs, I think those are charged credits for example.

**Comment by tylerjereddy** [comment]
CI, including wheel builds, is now fully green, with a 15 minute buffer added to the offending cirrus builds. May be worth circling back to that, but not right now. I'll put this in and try to get RC1 rolling soon-ish.

**Comment by andyfaff** [comment]
Are the windows wheels passing their tests because they're using numpy <2?

**Comment by tylerjereddy** [comment]
yes, that's right

**Pull request merged**